### PR TITLE
[beatport] Add new extractor

### LINKED ISSUE #5219: Add support for www.beatport.com
example url: https://www.beatport.com/flamingorecordingsbps/tracks/qpwhxukynk7r/burnin-up-original-mix

Thanks!

### PR DESCRIPTION
**Pull request by Thor77**
Fix #5219

### CODE REVIEW

File: youtube_dl/extractor/beatport.py [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+            'display_id': 'secrets-feat-vassy-original-mix',
+
+            'uploader': 'Musical Freedom',
+            'uploader_id': 'musicalfreedom',
+            'uploader_url': 'http://www.musicalfreedom.com',
+
+            'ext': 'mp4',
+            'title': 'Secrets feat. Vassy',
+            'duration': 267441,
+            'artist': ['Tiësto', 'VASSY', 'KSHMR'],
```
Comment (dstftw) [inline youtube_dl/extractor/beatport.py]: Must be string.

File: youtube_dl/extractor/beatport.py [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+            'genre': ['Electro House'],
```
Comment (dstftw) [inline youtube_dl/extractor/beatport.py]: Must be string.

File: youtube_dl/extractor/beatport.py [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+            'id': 'l2axnqcd6azu',
+            'display_id': 'secrets-feat-vassy-original-mix',
+
+            'uploader': 'Musical Freedom',
+            'uploader_id': 'musicalfreedom',
+            'uploader_url': 'http://www.musicalfreedom.com',
+
+            'ext': 'mp4',
+            'title': 'Secrets feat. Vassy',
+            'duration': 267441,
```
Comment (dstftw) [inline youtube_dl/extractor/beatport.py]: Must be in seconds.

File: youtube_dl/extractor/beatport.py (hunk 1 above)
Comment (Thor77) [inline youtube_dl/extractor/beatport.py]: Uh, is there no option to provide multiple artists(/genres)?

File: youtube_dl/extractor/beatport.py [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+            'duration': 272822,
+            'artist': ['Flatdisk']
+        }
+    }
+    ]
+
+    def _real_extract(self, url):
+        url_data = self._VALID_URL_RE.match(url)
+        api_base = 'https://apix.beatport.com'
+        api_data = self._download_json('{}/sounds/{}?format=json'.format(api_base, url_data.group('id')),
```
Comment (dstftw) [inline youtube_dl/extractor/beatport.py]: `{}` does not work on python 2.6.

File: youtube_dl/extractor/beatport.py [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+        return {
+            'url': url,
+            'id': url_data.group('id'),
+            'display_id': url_data.group('display_id'),
+
+            'uploader': api_data.get('owner_profile', {}).get('display_name'),
+            'uploader_url': api_data.get('owner_profile', {}).get('website_url'),
+            'uploader_id': url_data.group('uploader_id'),
+
+            'title': api_data.get('name'),
```
Comment (dstftw) [inline youtube_dl/extractor/beatport.py]: `title` is mandatory.

File: youtube_dl/extractor/beatport.py [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+            'duration': api_data.get('duration'),
+
+            'view_count': api_data.get('play_count'),
```
Comment (dstftw) [inline youtube_dl/extractor/beatport.py]: All integer fields must be wrapped in `utils.int_or_none`.

File: youtube_dl/extractor/beatport.py (hunk 1 above)
Comment (dstftw) [inline youtube_dl/extractor/beatport.py]: Multiple artists/genres should be provided as string.

Comment (Thor77) [inline youtube_dl/extractor/beatport.py]: I seperated them by `,` now.
During my research on that topic, i didn't really find any official information in id3's design-document. Some people said, it depends on the program to parse that information correctly, others suggested using `/`, `,` and `;`...

Comment (dstftw) [inline youtube_dl/extractor/beatport.py]: It should have the same format as on the particular website.

Comment (Thor77) [inline youtube_dl/extractor/beatport.py]: Are you reffering to this part of the specification?
```
All text information frames supports multiple strings, stored as a null
separated list, where null is reperesented by the termination code
for the charater encoding.
```
=> Use `\0`?

Comment (dstftw) [inline youtube_dl/extractor/beatport.py]: I'm not referring to any specification. It should be formatted the same way it formatted on beatport.com website.

Comment (Thor77) [inline youtube_dl/extractor/beatport.py]: [artist_list](https://www.beatport.com/musicalfreedom/tracks/l2axnqcd6azu/secrets-feat-vassy-original-mix)
Okey, i will add a space to the seperator.

File: youtube_dl/extractor/beatport.py [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+    def _real_extract(self, url):
+        url_data = self._VALID_URL_RE.match(url)
+        api_base = 'https://apix.beatport.com'
+        api_data = self._download_json('%s/sounds/%s?format=json' % (api_base, url_data.group('id')),
+                                       url_data.group('id'))
+        if api_data['is_streamable']:
+            url = self._download_json('%s/sounds/%s/stream?format=json' % (api_base, url_data.group('id')),
+                                      url_data.group('id'), note='Downloading JSON streamdata')['url']
+        else:
+            url = None
```
Comment (dstftw) [inline youtube_dl/extractor/beatport.py]: What is the point of this? If video url can't be extracted the whole further extraction makes no sense.

Comment (Thor77) [inline youtube_dl/extractor/beatport.py]: I don't want to execute foreseeable failing requests.
Additionally i didn't find a proper way to abort the extraction-process immediately (`InfoExtractor.report_warning` could've been useful here though) and so opted for this solution.

Comment (dstftw) [inline youtube_dl/extractor/beatport.py]: Raise `ExtractorError`.

File: youtube_dl/extractor/beatport.py [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+
+class BeatportIE(InfoExtractor):
+    _VALID_URL = r'https?://(?:www\.)?beatport\.com/(?P<uploader_id>\w+)/tracks/(?P<id>.+)/(?P<display_id>.+)'
+    _TESTS = [{
+        'url': 'https://www.beatport.com/musicalfreedom/tracks/l2axnqcd6azu/secrets-feat-vassy-original-mix',
+        'md5': '11e330f6236be07f61fcc4669e972cd8',
+        'info_dict': {
+            'id': 'l2axnqcd6azu',
+            'display_id': 'secrets-feat-vassy-original-mix',
+
```
Comment (dstftw) [inline youtube_dl/extractor/beatport.py]: Remove unnecessary blank lines.

File: youtube_dl/extractor/beatport.py [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+            'uploader_id': 'musicalfreedom',
+            'uploader_url': 'http://www.musicalfreedom.com',
+
+            'ext': 'mp4',
+            'title': 'Secrets feat. Vassy',
+            'duration': 267,
+            'artist': 'Tiësto, VASSY, KSHMR',
+            'genre': 'Electro House',
+            'release_year': '2015',
+            'thumbnail': 'https://api-media.beatport.com/sounds/l2axnqcd6azu/1431513641.jpg',
```
Comment (dstftw) [inline youtube_dl/extractor/beatport.py]: Replace with just `'thumbnail': 're:^https?://.*\.jpg',`.

File: youtube_dl/extractor/beatport.py [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+            'view_count': int,
+            'like_count': int
+        }
+    }
+    ]
+
+    def _real_extract(self, url):
+        url_data = self._VALID_URL_RE.match(url)
+        api_base = 'https://apix.beatport.com'
+        api_data = self._download_json('%s/sounds/%s?format=json' % (api_base, url_data.group('id')),
```
Comment (dstftw) [inline youtube_dl/extractor/beatport.py]: You are calculating `url_data.group('id')` over and over again. Extract it into a variable.

**Comment by dstftw** [comment]
Also squash commits when finished.

**Comment by Thor77** [comment]
Done. Thanks for your review!

**Comment by dstftw** [comment]
Apparently they are shutting down.
image
Does it work with pro.beatport.com?

**Comment by Thor77** [comment]
There's already an [extractor](https://github.com/rg3/youtube-dl/blob/master/youtube_dl/extractor/beatportpro.py) for BeatportPro

**Comment by Thor77** [comment]
I guess you can close this, as there's no sense to include an extractor for a closed platform.

**Comment by yan12125** [comment]
Now both https://www.beatport.com/musicalfreedom/tracks/l2axnqcd6azu/secrets-feat-vassy-original-mix and https://www.beatport.com/flamingorecordingsbps/tracks/qpwhxukynk7r/burnin-up-original-mix redirect to https://pro.beatport.com/. Closing.

**Pull request closed without merging**